### PR TITLE
[fix]: handle inputs with containers in mixed precision

### PR DESCRIPTION
**Pull request by min-xu-ai**
- this is an issue surfaces by vissl as well
- fix seems to be super simple
- also cleaned up two tests with respect to multiple such tests
  running back to back (they don't do that presently)

# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## What does this PR do?
Fixes # (issue).

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/nn/data_parallel/test_fsdp_input.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    try:
+        assert isinstance(fsdp_config, dict), str(fsdp_config)
+
+        class Model(Module):
+            def __init__(self):
+                super().__init__()
+                self.layer = Linear(4, 4)
+
+            def forward(self, input):
+                print("XXX", input)
```
Comment (min-xu-ai) [inline tests/nn/data_parallel/test_fsdp_input.py:50]: oops, will remove.

File: tests/nn/data_parallel/test_fsdp_grad_scaler.py [hunk 2]
```diff
@@ -17,28 +28,36 @@
...
-        output = model(input)
-        loss = F.mse_loss(input, output)
-
-    scaler.scale(loss).backward()
-    # TODO (Min): Need to fix. Details in issue #421.
-    with pytest.raises(RuntimeError):
-        scaler.step(optim)
-        scaler.update()
-    torch.distributed.destroy_process_group()
+    try:
```
Comment (min-xu-ai) [inline tests/nn/data_parallel/test_fsdp_grad_scaler.py:42]: no logically change in this and the wrap test. Just cleanup.

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py [hunk 3]
```diff
@@ -1145,15 +1139,14 @@ def assert_state(self, state: TrainingState) -> None:
...
     """
-    kwarg_keys, flat_args = pack_kwargs(*args, **kwargs)
-    tensor_inputs, packed_non_tensor_inputs = split_non_tensors(flat_args)
-    tensor_inputs = tuple(t.half() if torch.is_floating_point(t) else t for t in tensor_inputs)
-    flat_args = unpack_non_tensors(tensor_inputs, packed_non_tensor_inputs)
-    args, kwargs = unpack_kwargs(kwarg_keys, flat_args)
-    return args, kwargs
+
+    def fn(x: torch.Tensor) -> torch.Tensor:
+        if x.dtype is torch.float32:
```
Comment (myleott) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1145]: how about `torch.is_floating_point(x)`?

File: tests/nn/data_parallel/test_fsdp_grad_scaler.py [hunk 4]
```diff
@@ -17,28 +28,36 @@
...
+            loss = F.mse_loss(input, output)
+
+        scaler.scale(loss).backward()
+        # TODO (Min): Need to fix. Details in issue #421.
+        with pytest.raises(RuntimeError):
+            scaler.step(optim)
+            scaler.update()
+
+    finally:
+        # Clean-up is important or the next test in this file may fail to init the PG.
```
Comment (myleott) [inline tests/nn/data_parallel/test_fsdp_grad_scaler.py:60]: nice :)

**Review by myleott — APPROVED** [review]
This is awesome, much cleaner 😄

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py (hunk 3 above)
Comment (min-xu-ai) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1145]: I initially did that but then I think there is a slight help not calling half() if it is already half()? Am I missing it?

File: tests/nn/data_parallel/test_fsdp_grad_scaler.py (hunk 4 above)
Comment (min-xu-ai) [reply tests/nn/data_parallel/test_fsdp_grad_scaler.py:60]: Thanks for the review!

**Pull request merged**